File: 0907-koko-eating-bananas/0907-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        //sort the imput?
        int left = 1; // the minimum # of bnanas she can eat 
        int right = 1; // the max number she can eat; that has to be the maximum in the answer
        for(int pile : piles){
            right = Math.max(right, pile);
        }
        // So, what's different here is we are running binary search, not on our array but our solution space
        // which is between 1 and right
        while(left < right){
            int mid = left + (right - left)/2;
            //check for mid if its possible
            int hours = 0;
            for(int pile: piles){
                hours += Math.ceil((double)pile / mid);
            }
            if(hours > h){ //meaning not possible, answer has to be greatre than mid
                left = mid + 1;
            }else{
                right = mid ;
            }
        }
        return left;

    }
}

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.rs
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let max_possible = piles.iter().max().unwrap();
        let mut left = 1;
        let mut right = *max_possible;

        if piles.len() == h as usize{
            return *max_possible;
        }
        while left < right{
            let mid = left + (right - left) / 2;
            if Self::check(&piles, mid, h){
                println!("{}", mid);
                right = mid;
            } else{
                left = mid + 1;
            }
            
        }
        left
    }
    fn check(piles: &[i32], k : i32, h:i32) -> bool{
        let mut curr = 0;
        for pile in piles.iter(){
            curr += (pile + k - 1)/k;
        }
        if curr <= h{
            return true;
        }
        false
    }
}